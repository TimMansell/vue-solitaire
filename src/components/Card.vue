<template>
  <div
    class="card"
    :class="classes"
    @click="selectCard($event, id)"
    :data-test="`card-${value}${suit}`">
    <SvgIcon
      v-if="visible"
      :name="`${this.value}${this.suit.toUpperCase()}`" />

    <SvgIcon
      v-if="!visible"
      name="Card_back_17" />
  </div>
</template>

<script>
import SvgIcon from '@/components/SvgIcon.vue';

export default {
  name: 'Card',
  components: {
    SvgIcon,
  },
  props: {
    id: {
      type: [String, Number],
      default: 0,
    },
    value: {
      type: String,
      default: 'A',
    },
    suit: {
      type: String,
      default: 'c',
    },
    order: {
      type: Number,
      default: 0,
    },
    position: {
      type: Array,
      default: () => [],
    },
    visible: {
      type: Boolean,
      default: false,
    },
    revealed: {
      type: Boolean,
      default: false,
    },
    clickable: {
      type: Boolean,
      default: true,
    },
  },
  computed: {
    classes() {
      const { selectedCardId } = this.$store.getters;

      return {
        'card--is-visible': this.visible,
        'card--is-s': this.suit === 's',
        'card--is-d': this.suit === 'd',
        'card--is-h': this.suit === 'h',
        'card--is-c': this.suit === 'c',
        'card--is-selected': selectedCardId === this.id,
        'card--is-not-clickable': !this.clickable,
      };
    },
  },
  methods: {
    selectCard(e, id) {
      const { selectedCardId } = this.$store.getters;

      if (!selectedCardId) {
        e.stopPropagation();

        if (this.clickable && this.visible) {
          this.$store.dispatch('selectCard', id);
        }
      }
    },
  },
};
</script>

<style scoped lang="scss">
.card {
  transition: all .05s ease-in-out;
  transform-style: preserve-3d;

  &:nth-of-type(n+2) {
    margin-top: -#{$card-height * $card-spacer};

    @media (min-width: $bp-sm) {
      margin-top: -#{$card-height-sm * $card-spacer};
    }

    @media (min-width: $bp-md) {
      margin-top: -#{$card-height-md * $card-spacer};
    }

    @media (min-width: $bp-lg) {
      margin-top: -#{$card-height-lg * $card-spacer};
    }

    @media (min-width: $bp-xl) {
      margin-top: -#{$card-height-xl * $card-spacer};
    }
  }

  &--is-s {
    &--color-blind {
      color: black;
    }
  }

  &--is-d {
    &--color-blind {
      color: blue;
    }
  }

  &--is-h {
    &--color-blind {
      color: red;
    }
  }

  &--is-c {
    &--color-blind {
      color: orange;
    }
  }

  &--is-selected {
    transform: scale(1.1);
  }

  &--is-not-clickable {
    pointer-events:none;
  }
}
</style>
