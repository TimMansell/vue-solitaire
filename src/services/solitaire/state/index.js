const setState = (state, newState) => ({ ...state, ...newState });

export default setState;
