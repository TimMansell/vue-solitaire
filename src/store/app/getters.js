const getters = {
  game: ({ game }) => game,
  isGameLoading: ({ isGameLoading }) => isGameLoading,
  isGamePaused: ({ isGamePaused }) => isGamePaused,
  isOverlayVisible: ({ isOverlayVisible }) => isOverlayVisible,
  gameOutcome: ({ gameOutcome }) => gameOutcome,
  timer: ({ game }) => game.time,
  moves: ({ game }) => game.moves.length,
  hasGameUpdated: ({ hasGameUpdated }) => hasGameUpdated,
  isLatestVersion: ({ isLatestVersion }) => isLatestVersion,
  version: ({ version }) => version,
  isConnecting: ({ connection }) => connection.isConnecting,
  isOnline: ({ connection }) => connection.isOnline,
  isCompletedGame: ({ gameOutcome }) =>
    gameOutcome.hasGameLost || gameOutcome.hasGameWon,
};

export default getters;
