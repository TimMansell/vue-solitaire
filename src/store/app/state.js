// Return state as function so we can reset it.
const state = () => ({
  game: {
    moves: [],
    time: 0,
  },
  gameOutcome: {
    hasGameWon: false,
    hasGameLost: false,
  },
  isGameLoading: true,
  isGamePaused: false,
  isOverlayVisible: true,
  versionMatch: true,
});

export default state;
