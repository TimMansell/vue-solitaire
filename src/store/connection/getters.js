const getters = {
  isConnecting: ({ isConnecting }) => isConnecting,
  hasConnectionError: ({ isConnecting, isOnline }) =>
    !isOnline && !isConnecting,
};

export default getters;
