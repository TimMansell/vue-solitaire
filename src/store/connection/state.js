const state = () => ({
  isConnecting: false,
  hasConnected: false,
  isOnline: false,
});

export default state;
