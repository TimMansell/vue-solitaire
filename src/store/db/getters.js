const getters = {
  userStats: (state) => state.userStats,
  game: (state) => state.game,
};

export default getters;
