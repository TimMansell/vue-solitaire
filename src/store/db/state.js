// Return state as function so we can reset it.
const state = () => ({
  userStats: {
    totalGames: null,
  },
  game: {
    id: null,
    moves: 0,
  },
});

export default state;
