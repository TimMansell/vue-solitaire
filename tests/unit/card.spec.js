import { shallowMount, createLocalVue } from '@vue/test-utils';
import Vuex from 'vuex';
import Card from '@/components/Card.vue';
import state from '@/store/state';
import getters from '@/store/getters';

const localVue = createLocalVue();

localVue.use(Vuex);

describe('Card.vue', () => {
  let store;

  const actions = {
    selectCard: jest.fn(),
  };

  beforeEach(() => {
    store = new Vuex.Store({
      state,
      getters,
      actions,
    });
  });

  it('matches snapshot', () => {
    const wrapper = shallowMount(Card, { store, localVue });

    expect(wrapper).toMatchSnapshot();
  });

  it('should have correct props', () => {
    const wrapper = shallowMount(Card, {
      store,
      localVue,
      propsData: {
        value: 'K',
        suit: 'd',
        order: 1,
        visible: true,
        clickable: false,
      },
    });

    expect(wrapper.props().value).toBe('K');
    expect(wrapper.props().suit).toBe('d');
    expect(wrapper.props().order).toBe(1);
    expect(wrapper.props().visible).toBe(true);
    expect(wrapper.props().clickable).toBe(false);
  });

  it('should render a diamond card', () => {
    const wrapper = shallowMount(Card, {
      store,
      localVue,
      propsData: {
        value: '6',
        suit: 'd',
      },
    });

    expect(wrapper.classes()).toContain('card--is-d');
  });

  it('should render a club card', () => {
    const wrapper = shallowMount(Card, {
      store,
      localVue,
      propsData: {
        value: '6',
        suit: 'c',
      },
    });

    expect(wrapper.classes()).toContain('card--is-c');
  });

  it('should render a heart card', () => {
    const wrapper = shallowMount(Card, {
      store,
      localVue,
      propsData: {
        value: '6',
        suit: 'h',
      },
    });

    expect(wrapper.classes()).toContain('card--is-h');
  });

  it('should render a spade card', () => {
    const wrapper = shallowMount(Card, {
      store,
      localVue,
      propsData: {
        value: '6',
        suit: 's',
      },
    });

    expect(wrapper.classes()).toContain('card--is-s');
  });

  it('should render a hidden card', () => {
    const wrapper = shallowMount(Card, {
      store,
      localVue,
      propsData: {
        value: '6',
        suit: 's',
      },
    });

    expect(wrapper.classes('card--is-visible')).toBe(false);
  });

  it('should render a visible card', () => {
    const wrapper = shallowMount(Card, {
      store,
      localVue,
      propsData: {
        visible: true,
      },
    });

    expect(wrapper.classes()).toContain('card--is-visible');
  });

  it('should render a selected card', () => {
    const newGetters = {
      selectedCardId: () => 0,
    };

    const storeSelected = new Vuex.Store({
      state,
      getters: newGetters,
      actions,
    });

    const wrapper = shallowMount(Card, {
      store: storeSelected,
      localVue,
      propsData: {
        value: 'K',
        suit: 'c',
      },
    });

    expect(wrapper.classes()).toContain('card--is-selected');
  });

  it('should not call store action "selectCard" when clicked', () => {
    const wrapper = shallowMount(Card, {
      store,
      localVue,
      propsData: {
        clickable: false,
      },
    });

    wrapper.find('[data-test="card-Ac"]').trigger('click');

    expect(wrapper.classes()).toContain('card--is-not-clickable');
    expect(actions.selectCard).not.toHaveBeenCalled();
  });

  it('should not call store action "selectCard" when clicked when not visible', () => {
    const wrapper = shallowMount(Card, { store, localVue });

    wrapper.find('[data-test="card-Ac"]').trigger('click');

    expect(actions.selectCard).not.toHaveBeenCalled();
  });

  it('should call store action "selectCard" when clicked when visible', () => {
    const wrapper = shallowMount(Card, {
      store,
      localVue,
      propsData: {
        visible: true,
      },
    });

    wrapper.find('[data-test="card-Ac"]').trigger('click');

    expect(actions.selectCard).toHaveBeenCalled();
  });
});
