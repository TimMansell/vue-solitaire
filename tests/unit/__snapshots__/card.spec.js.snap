// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card.vue matches snapshot 1`] = `
<div data-test="card-Ac" class="card card--is-c">
  <!---->
</div>
`;
